### PR TITLE
handling deployvm API

### PR DESCRIPTION
**Pull request by shwstppr**
Signed-off-by: Abhishek Kumar <abhishek.mrt22@gmail.com>

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by shwstppr** [comment]
@DaanHoogland please see

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 1]
```diff
@@ -373,6 +381,13 @@ public Long getZoneId() {
     }
 
     public List<Long> getNetworkIds() {
+        if (MapUtils.isNotEmpty(vAppNetworks)) {
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:408]: why is this condition repeated in the private method? is there a chance it has changed?

Comment (shwstppr) [reply api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:408]: I was not sure if the private method will be used only in this section. It won't be changed. Will refactor it once we have sync with Paul on mapping @DaanHoogland

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -5198,6 +5199,35 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
...
+                for (NetworkPrerequisiteTO networkPrerequisiteTO : networkPrerequisiteTOList) {
+                    Long networkId = networkMap.get(networkPrerequisiteTO.getInstanceID());
+                    if (networkId == null && networkPrerequisiteTO.isAutomaticAllocation()) {
+                        networkId = defaultNetwork.getId();
+                    }
+                    networkIds.add(networkId);
+                }
+            }
+        }
+
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: @DaanHoogland this is what I proposed. API will supply a list of networkIds with same size as the number of NICs in the template. Though some elements in this list can be null when user has not passed any mapping and automaticAllocation is false.

Now the problem with this is, at some point during deploy we check if the network/network id in the list is not null. That will need some refactoring. Also, refactoring will be needed while creating NicProfiles https://github.com/shapeblue/cloudstack/blob/599b43a7ca08f55e6becb991ad0541647f318373/server/src/main/java/com/cloud/vm/UserVmManagerImpl.java#L3630-L3705

**Comment by nvazquez** [comment]
Thanks @shwstppr, looking good. Let me know once its ready so we incorporate it into the big PR

**Review by nvazquez — APPROVED** [review]

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 3]
```diff
@@ -324,6 +332,30 @@ public Long getDomainId() {
         return map;
     }
 
+    public Map<Integer, Long> getVmNetworkMap() {
+        Map<Integer, Long> map = new HashMap<>();
+        Collection parameterCollection = vAppNetworks.values();
```
Comment (nvazquez) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:337]: Im getting an NPE at this line, used the old UI can you check @shwstppr?:

````
ERROR [c.c.a.ApiServer] (qtp1163480574-33:ctx-ab3cbe5e ctx-277571f8) (logid:9e625094) unhandled exception executing api command: [Ljava.lang.String;@49d2e4b1
java.lang.NullPointerException
	at org.apache.cloudstack.api.command.user.vm.DeployVMCmd.getVmNetworkMap(DeployVMCmd.java:337)
	at com.cloud.vm.UserVmManagerImpl.createVirtualMachine(UserVmManagerImpl.java:5205)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
````

**Pull request merged**